common/static/common/js/discussion/utils.js

```diff
@@ -353,84 +356,107 @@
+                );
             });
-            text = $div.html();
-            return text;
+            return edx.HtmlUtils.HTML($div.html());
         };
 
-        DiscussionUtil.unescapeHighlightTag = function(text) {
-            return text.replace(
-                /\&lt\;highlight\&gt\;/g,
-                "<span class='search-highlight'>").replace(/\&lt\;\/highlight\&gt\;/g, "</span>"
+        DiscussionUtil.unescapeHighlightTag = function(htmlSnippet) {
+            return edx.HtmlUtils.HTML(
+                htmlSnippet.toString().replace(
+                    /\&lt\;highlight\&gt\;/g,
+                    "<span class='search-highlight'>").replace(/\&lt\;\/highlight\&gt\;/g, "</span>"
+                )
             );
         };
 
-        DiscussionUtil.stripHighlight = function(text) {
-            return text.replace(
-                /\&(amp\;)?lt\;highlight\&(amp\;)?gt\;/g, "").replace(/\&(amp\;)?lt\;\/highlight\&(amp\;)?gt\;/g, ""
-            );
+        DiscussionUtil.stripHighlight = function(htmlString) {
+            return htmlString
+                    .replace(/\&(amp\;)?lt\;highlight\&(amp\;)?gt\;/g, "")
+                    .replace(/\&(amp\;)?lt\;\/highlight\&(amp\;)?gt\;/g, "");
         };
 
-        DiscussionUtil.stripLatexHighlight = function(text) {
-            return this.processEachMathAndCode(text, this.stripHighlight);
+        DiscussionUtil.stripLatexHighlight = function(htmlSnippet) {
+            return this.processEachMathAndCode(htmlSnippet, this.stripHighlight);
         };
 
-        DiscussionUtil.markdownWithHighlight = function(text) {
+        /**
+         * Processes markdown into formatted text and handles highlighting.
+         * @param unsafeText - raw markdown text, with all HTML entitites being *unescaped*.
+         * @returns HtmlSnippet
+         */
+        DiscussionUtil.markdownWithHighlight = function(unsafeText) {
             var converter;
-            text = text.replace(/^\&gt\;/gm, ">");
+            unsafeText = unsafeText.replace(/^\&gt\;/gm, ">");
             converter = Markdown.getMathCompatibleConverter();
-            text = this.unescapeHighlightTag(this.stripLatexHighlight(converter.makeHtml(text)));
-            return text.replace(/^>/gm, "&gt;");
+            /*
+            converter.makeHtml have at least two issues:
+            * TODO: converter.makeHtml is not HtmlSnippet aware, so we must pass unescaped text raw text
+            * TODO: converter.makeHtml strips html tags in post body, but escapes in code blocks
+            *       as a result we can't preserve tags in body and have properly escaped tags in code blocks:
+            *       either body tags are not escaped, or code tags are double escaped, unless we specifically unescape
+            *       html tags in code templates and pass escaped code here.
+            *       Current implementation does not do that and just strips tags in body (looks like it is by design)
+            */
```

Comment:
@e-kolpakov: Can we work through this comment here and you can update when you squash, assuming you don't want to run tests again and again?  There is a typo, and I think I would comment this differently.

What do you think of:
```
           /*
            * converter.makeHtml and HTML escaping:
            * - converter.makeHtml is not HtmlSnippet aware, so we must pass unescaped text raw text
            * - converter.makeHtml strips html tags in post body and escapes in code blocks by design.
            *    HTML tags are not supported.  Only markdown is supported.
            */
```

Is this accurate?  I'm not sure why there is any need for a TODO here.